site/src/components/Resources/PortForwardButton.tsx

```diff
@@ -0,0 +1,274 @@
+export const PortForwardButton: React.FC<PortForwardButtonProps> = (props) => {
+  const anchorRef = useRef<HTMLButtonElement>(null)
+  const [isOpen, setIsOpen] = useState(false)
+  const id = isOpen ? "schedule-popover" : undefined
+  const styles = useStyles()
+  const { data: listeningPorts } = useQuery({
+    queryKey: ["portForward", props.agentId],
+    queryFn: () => getAgentListeningPorts(props.agentId),
+    initialData: MockListeningPortsResponse,
+  })
+
+  const onClose = () => {
+    setIsOpen(false)
+  }
+
+  return (
+    <>
+      <SecondaryAgentButton
+        disabled={!listeningPorts}
+        ref={anchorRef}
+        onClick={() => {
+          setIsOpen(true)
+        }}
+      >
+        Ports
+        {listeningPorts ? (
+          <Box
+            sx={{
+              fontSize: 12,
+              fontWeight: 500,
+              height: 16,
+              padding: (theme) => theme.spacing(0, 1),
+              borderRadius: 7,
+              display: "flex",
+              alignItems: "center",
+              justifyContent: "center",
+              backgroundColor: colors.gray[11],
+              ml: 1,
+            }}
+          >
+            {listeningPorts.ports.length}
+          </Box>
+        ) : (
+          <CircularProgress size={10} sx={{ ml: 1 }} />
+        )}
+      </SecondaryAgentButton>
+      <Popover
+        classes={{ paper: styles.popoverPaper }}
+        id={id}
+        open={isOpen}
+        anchorEl={anchorRef.current}
+        onClose={onClose}
+        anchorOrigin={{
+          vertical: "bottom",
+          horizontal: "left",
+        }}
+        transformOrigin={{
+          vertical: "top",
+          horizontal: "left",
+        }}
+      >
+        <HelpTooltipTitle>Port forward</HelpTooltipTitle>
+        <PortForwardPopoverView {...props} ports={listeningPorts?.ports} />
+      </Popover>
+    </>
+  )
+}
+
+export const PortForwardPopoverView: React.FC<
+  PortForwardButtonProps & { ports?: WorkspaceAgentListeningPort[] }
+> = (props) => {
+  const { host, workspaceName, agentName, username, ports } = props
+
+  return (
+    <>
+      <Box
+        sx={{
+          padding: (theme) => theme.spacing(2.5),
+          borderBottom: (theme) => `1px solid ${theme.palette.divider}`,
+        }}
+      >
+        <HelpTooltipTitle>Ports</HelpTooltipTitle>
+        <HelpTooltipText
+          sx={{ color: (theme) => theme.palette.text.secondary }}
+        >
+          {ports?.length === 0
+            ? "No open ports were detected."
+            : "Port URLs are only accessible by you."}
+        </HelpTooltipText>
+        <Box sx={{ marginTop: (theme) => theme.spacing(1.5) }}>
+          {ports?.map((p) => {
+            const url = portForwardURL(
+              host,
+              p.port,
+              agentName,
+              workspaceName,
+              username,
+            )
+            const label =
+              p.process_name !== "" ? p.process_name : `Port ${p.port}`
```

Comment:
```suggestion
              p.process_name !== "" ? p.process_name : `${p.port}`
```

Let's just re-render the port #